request-manager/graph/resolver.go

```diff
@@ -121,193 +123,168 @@ func (o *resolver) RequestArgs(jobArgs map[string]interface{}) ([]proto.RequestA
 	}
 
-	// Traverse sequence graph in topological order. Each sequence graph node corresponds to either a single
-	// job node, or an entire subgraph of nodes. Build the subgraph and insert it into the larger sequence
-	// graph.
+	// Traverse sequence graph in topological order. Each sequence graph node
+	// corresponds to either a single job node, or an entire subgraph of nodes.
+	// Build the subgraph and insert it into the larger sequence graph.
 	idMap := map[string]*Graph{} // sequence graph node id --> corresponding subgraph
 	for _, seqGraphNode := range seqGraph.Order {
-		n := seqGraphNode.Spec
-
-		// Find out how many times this node has to be repeated
-		elements, lists, err := o.getIterators(n, jobArgs)
+		nodeSpec := seqGraphNode.Spec
+
+		// 1. Do sequence expansion.
+		// Find out how many times this node has to be repeated.
+		// If this node has a `each:` field, then lists[i] and
+		// elements[i] describe the ith `list:element`. Specifically,
+		// lists[i] is the actual list value, not just the name of the
+		// job arg.
+		elements, lists, err := r.getIterators(nodeSpec, jobArgs)
 		if err != nil {
-			return nil, fmt.Errorf("in seq %s, node %s: invalid 'each:' %s", sequenceName, n.Name, err)
+			return nil, fmt.Errorf("in seq %s, node %s: invalid 'each:' %s", sequenceName, nodeSpec.Name, err)
 		}
 
-		// All the graphs that make up this sequence graph node's subgraph
+		// All the graphs that make up this sequence graph node's subgraph,
+		// one for each time the node is repeated
 		components := []*Graph{}
 
 		// If no repetition is needed, this loop will only execute once
+		// with a dummy `each:` entry `[""]:nil`.
 		for i, _ := range lists[0] {
-			// Copy the required args into a separate args map here.
-			// Do the necessary remapping here.
-			jobArgsCopy, err := remapNodeArgs(n, jobArgs)
+			// Copy the required args into a separate args map here
+			// and do the necessary remapping
+			jobArgsCopy, err := remapNodeArgs(nodeSpec, jobArgs)
 			if err != nil {
 				return nil, err
 			}
 
-			// Add the element to the node args unless there is none for this node
+			// Given "each: foos:foo", we add jobArgs[foo] = jobArgs[foos][i],
+			// unless this is the dummy `each:` that ensures that we
+			// run this loop exactly once, in which case elt == nil.
 			for j, elt := range elements {
-				if elt != "" {
+				if elt != nil {
 					// This won't panic because we have earlier asserted that
 					// len(elements) == len(lists)
-					jobArgsCopy[elt] = lists[j][i]
+					jobArgsCopy[*elt] = lists[j][i]
 				}
 			}
 
-			// Add the "if" to the node args if it's present
-			if n.If != nil {
-				if ifArg, ok := jobArgs[*n.If]; ok {
-					jobArgsCopy[*n.If] = ifArg
+			// If this is a conditional node, add the "if" job arg.
+			// Static checks asserted if != nil for conditional nodes.
+			if nodeSpec.IsConditional() {
+				if ifArg, ok := jobArgs[*nodeSpec.If]; ok {
+					jobArgsCopy[*nodeSpec.If] = ifArg
 				}
```

Comment:
The inner `if {` block can go in the next block which starts with the same check: `if nodeSpec.IsConditional() {`.

Reply:
The code is slightly messier that way; right now, it does all the job arg stuff (starts with the `jobArgsCopy` stuff, ends with this block), then starts resolving the node. Moving the `if` block in with the next one mixes job arg stuff with resolving.

Reply:
I see. Then I'd sell this as "separation of concerns" rather than clean/messy code. In essence we have,
```go
if X {
  // logic 1
}

if X {
  // logic 2
}
```
which raises a flag wrt another very valuable principle: DRY (Don't Repeat Yourself). An obvious duplication like that is messy code, imho, and a potential for bugs because engineers can forget to update the logic in both places (and it also raises questions about when, wrt `X`, we update logic 1, 2, or both).

But I see that you want to have all job arg logic together at top of loop, and all resolver logic together at bottom of loop. Separation of concerns is good, too. In other code, we might do that better by sticking all the job arg logic is a single func (or some other method of delineation). But this loop is small enough to keep it all in the loop. So to achieve virtual separation of concerns, i.e. signal to FE that we're going to duplicate code not by accident or oversight:

```go
		for i, _ := range lists[0] {
			// ---------------------------------------------------------------
			// Copy and prepare job args for this graph node

			...

			// ---------------------------------------------------------------
			// Resolve this graph node into a request subgarph

			var reqSubgraph *Graph
			if nodeSpec.IsConditional() {
```
Then the FE who's hopefully reading these should realize: job arg logic up here, and resolver logic down there, and maybe there's some overlap but that's ok because the separation is by design.